File: src/bot/listeners/contents/photoHandler.ts
import { format } from 'date-fns';
import { Client } from 'tdl';
import { message, updateNewMessage } from 'tdlib-types';
import { PHOTO_DESTINATION } from '../../../constants';
import { isSender } from '../../filters';
import { debug, refreshChats } from '../../helper';

export const photoHandler = async (
  update: updateNewMessage,
  client: Client
): Promise<void> => {
  // only accept the exact sender
  if (!isSender(update)) {
    debug('Received update from other sender', JSON.stringify(update));
    return;
  }

  console.log('Received new photo');

  // get chats
  const { topic } = await refreshChats(client, update.message.chat_id, {
    chatId: PHOTO_DESTINATION.chat,
    topicName: PHOTO_DESTINATION.thread,
  });

  // forward photo
  const messages = await client.invoke({
    _: 'forwardMessages',
    chat_id: PHOTO_DESTINATION.chat,
    message_thread_id: topic?.info.message_thread_id,
    message_ids: [update.message.id],
    from_chat_id: update.message.chat_id,
  });

  const message = messages.messages[0] as message;

  // send timestamp
  const timestamp = format(
    new Date(update.message.date * 1000),
    'yyyy-MM-dd HH:mm X'
  );
  await client.invoke({
    _: 'sendMessage',
    chat_id: message.chat_id,
    message_thread_id: message.message_thread_id,
    reply_to: { _: 'inputMessageReplyToMessage', message_id: message.id },
    input_message_content: {
      _: 'inputMessageText',
      text: { _: 'formattedText', text: timestamp },
    },
  });
};

File: src/main/java/com/pirasalbe/services/telegram/listeners/contents/PhotoHandler.java
package com.pirasalbe.services.telegram.listeners.contents;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.function.BiConsumer;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import com.pirasalbe.configurations.TelegramConfiguration;
import com.pirasalbe.helpers.DateHelper;
import com.pirasalbe.helpers.FilterHelper;
import com.pirasalbe.helpers.TelegramBotHelper;
import com.pirasalbe.helpers.models.ForumInfo;
import com.pirasalbe.helpers.models.RefreshResult;

import it.tdlight.client.Result;
import it.tdlight.client.SimpleTelegramClient;
import it.tdlight.jni.TdApi.FormattedText;
import it.tdlight.jni.TdApi.ForwardMessages;
import it.tdlight.jni.TdApi.InputMessageReplyToMessage;
import it.tdlight.jni.TdApi.InputMessageText;
import it.tdlight.jni.TdApi.Message;
import it.tdlight.jni.TdApi.Messages;
import it.tdlight.jni.TdApi.SendMessage;
import it.tdlight.jni.TdApi.TextEntity;
import it.tdlight.jni.TdApi.UpdateNewMessage;

@Component
public class PhotoHandler implements BiConsumer<SimpleTelegramClient, UpdateNewMessage> {

	private static final Logger LOGGER = LoggerFactory.getLogger(PhotoHandler.class);

	@Autowired
	private TelegramConfiguration configuration;

	@Override
	public void accept(SimpleTelegramClient client, UpdateNewMessage update) {
		if (!FilterHelper.isSender(configuration.getChatId(), update)) {
			LOGGER.debug("Received update from other sender", update);
			return;
		}

		LOGGER.info("Received new photo");

		// get chats
		RefreshResult refreshResult = TelegramBotHelper.refreshChats(client, update.message.chatId, new ForumInfo(
				configuration.getPhotoDestinationChatId(), configuration.getPhotoDestinationThreadName()));

		Result<Messages> messages = TelegramBotHelper.sendSync(client,
				new ForwardMessages(configuration.getPhotoDestinationChatId(),
						refreshResult.getTopic().map(topic -> topic.info.messageThreadId).orElse(0l),
						update.message.chatId, new long[] { update.message.id }, null, false, false));

		Message message = messages.get().messages[0];

		LocalDateTime timestamp = DateHelper.integerToLocalDateTime(update.message.date);
		String text = timestamp.format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm"));

		TelegramBotHelper.sendSync(client,
				new SendMessage(message.chatId, message.messageThreadId,
						new InputMessageReplyToMessage(message.chatId, message.id, null), null, null,
						new InputMessageText(new FormattedText(text, new TextEntity[] {}), null, false)));
	}

}
